api/handlers/reporthandler.py

```diff
@@ -0,0 +1,257 @@
+import json
+import bson
+import dateutil
+import copy
+
+from .. import base
+from .. import config
+from .. import util
+
+log = config.log
+
+EIGHTEEN_YEARS_IN_SEC = 568036800
+
+class APIReportException(Exception):
+    pass
+
+class ReportHandler(base.RequestHandler):
+
+    def __init__(self, request=None, response=None):
+        super(ReportHandler, self).__init__(request, response)
+
+    def get(self, report_type, **kwargs):
+        report = None
+        if report_type == 'site':
+            report = SiteReport()
+
+        elif report_type == 'project':
+            project_list = self.request.GET.getall('projects')
+            start_date = self.get_param('start_date')
+            end_date = self.get_param('end_date')
+
+            if len(project_list) < 1:
+                self.abort(400, 'List of projects requried for Project Report')
+            if start_date is not None:
+                start_date = dateutil.parser.parse(self.get_param('start_date'))
+            if end_date is not None:
+                end_date = dateutil.parser.parse(self.get_param('end_date'))
+
+            report = ProjectReport(map(bson.ObjectId, project_list),
+                                   start_date=start_date,
+                                   end_date=end_date)
+
+        else:
+            # They should never even get this far because of filtering in api.py
+            self.abort(400, 'The report type {} is not supported'.format(report_type))
+
+        if self.superuser_request or report.user_can_generate(self.uid):
+            return report.build()
+        else:
+            self.abort(403, 'User {} does not have required permissions to generate report'.format(self.uid))
+
+
+def _get_result(output):
+    """
+    Helper function for extracting mongo aggregation results
+
+    Given the output of a mongo aggregation call, checks 'ok' field
+    If not 1.0 or 'result' field does not exist, throws APIReportException
+    """
+
+    if output.get('ok', 0) is not 1.0:
+        result = output.get('result')
+        if result is not None:
+            return result[0]
+
+    raise APIReportException
+
+class Report(object):
+
+    def user_can_generate(self, uid):
```

Comment:
This looks more like an interface than an abstract class right now but future mentioned functionality like generating a csv from the json data would go here as a common method. Or it could be handled at the handler level. Curious see hear any thoughts on designing this module.